cpp/include/raft/core/csr_matrix.hpp

```diff
@@ -0,0 +1,307 @@
+                              compressed_structure_view<IndptrType, IndicesType, NZType, is_device>,
+                              is_device> {
+ public:
+  csr_matrix_view(
+    raft::span<ElementType, is_device> element_span,
+    compressed_structure_view<IndptrType, IndicesType, NZType, is_device> structure_view)
+    : sparse_matrix_view<ElementType,
+                         compressed_structure_view<IndptrType, IndicesType, NZType, is_device>,
+                         is_device>(element_span, structure_view){};
+};
+
+template <typename ElementType,
+          typename IndptrType,
+          typename IndicesType,
+          typename NZType,
+          bool is_device,
+          template <typename T>
+          typename ContainerPolicy,
+          SparsityType sparsity_type = SparsityType::OWNING,
+          typename structure_type    = std::conditional_t<
+            sparsity_type == SparsityType::OWNING,
+            compressed_structure<IndptrType, IndicesType, NZType, is_device, ContainerPolicy>,
+            compressed_structure_view<IndptrType, IndicesType, NZType, is_device>>>
+class csr_matrix
+  : public sparse_matrix<ElementType,
+                         structure_type,
+                         csr_matrix_view<ElementType, IndptrType, IndicesType, NZType, is_device>,
+                         is_device,
+                         ContainerPolicy> {
+ public:
+  using element_type        = ElementType;
+  using structure_view_type = typename structure_type::view_type;
+  static constexpr auto get_sparsity_type() { return sparsity_type; }
+  using sparse_matrix_type =
+    sparse_matrix<ElementType,
+                  structure_type,
+                  csr_matrix_view<ElementType, IndptrType, IndicesType, NZType, is_device>,
+                  is_device,
+                  ContainerPolicy>;
+  using container_type = typename ContainerPolicy<ElementType>::container_type;
+
+  template <typename = typename std::enable_if<sparsity_type == SparsityType::OWNING>>
+  csr_matrix(raft::resources const& handle,
+             IndptrType n_rows,
+             IndicesType n_cols,
+             NZType nnz = 0) noexcept(std::is_nothrow_default_constructible_v<container_type>)
+    : sparse_matrix_type(handle, n_rows, n_cols, nnz){};
+
+  // Constructor that owns the data but not the structure
+
+  template <typename = typename std::enable_if<sparsity_type == SparsityType::PRESERVING>>
+  csr_matrix(raft::resources const& handle, std::shared_ptr<structure_type> structure) noexcept(
+    std::is_nothrow_default_constructible_v<container_type>)
+    : sparse_matrix_type(handle, structure){};
+
+  /**
+   * Initialize the sparsity on this instance if it was not known upon construction
+   * Please note this will resize the underlying memory buffers
+   * @param nnz new sparsity to initialize.
+   */
+  template <typename = std::enable_if<sparsity_type == SparsityType::OWNING>>
+  void initialize_sparsity(NZType nnz)
```

Comment:
Same question as before. This function is compile-time hidden if you have `SparsityType::PRESERVING`. However, if you call `csr_matrix<..., SparsityType::PRESERVING, ...>::structure_view().initialize_sparsity()` then it's a runtime error now